syntax = "proto3";

// API Version 1.0 (RN-007: Versionamento de API)
package chat4all.v1;

option java_multiple_files = true;
option java_package = "chat4all.grpc.generated.v1";
option java_outer_classname = "FilesProto";

// Serviço de Arquivos
service FileService {
  // Upload de arquivo (streaming com suporte a resumo)
  rpc UploadFile(stream FileChunk) returns (UploadFileResponse);
  
  // Download de arquivo (streaming)
  rpc DownloadFile(DownloadFileRequest) returns (stream FileChunk);
  
  // Obter metadados do arquivo
  rpc GetFileMetadata(GetFileMetadataRequest) returns (FileMetadata);
  
  // Retomar upload interrompido (RF-004)
  rpc ResumeUpload(ResumeUploadRequest) returns (ResumeUploadResponse);
}

message FileChunk {
  bytes content = 1;
  FileMetadata metadata = 2; // Enviado apenas no primeiro chunk
  int64 offset = 3;           // Byte offset para upload resumível
  string session_id = 4;      // ID da sessão para retomar upload
  string chunk_checksum = 5;  // SHA-256 do chunk (validação de integridade)
}

message FileMetadata {
  string file_id = 1;
  string filename = 2;
  int64 size_bytes = 3;
  string mime_type = 4;
  string checksum = 5;
  string conversation_id = 6;
  int64 uploaded_at = 7;
}

message UploadFileResponse {
  string file_id = 1;
  string filename = 2;
  int64 size_bytes = 3;
  string checksum = 4;
  string storage_path = 5;
}

message DownloadFileRequest {
  string file_id = 1;
}

message GetFileMetadataRequest {
  string file_id = 1;
}

message ResumeUploadRequest {
  string session_id = 1;
}

message ResumeUploadResponse {
  string session_id = 1;
  int64 bytes_uploaded = 2;  // Quantos bytes já foram recebidos
  bool can_resume = 3;        // Se a sessão ainda é válida
}

